packages/annotations/src/vote.ts

```diff
@@ -0,0 +1,125 @@
+ * Apache-2.0 */
+
+import {
+  ArcGISRequestError,
+  ArcGISAuthError,
+  IRequestOptions
+} from "@esri/arcgis-rest-request";
+
+import {
+  addFeatures,
+  deleteFeatures,
+  queryFeatures,
+  IAddFeaturesRequestOptions,
+  IDeleteFeaturesRequestOptions,
+  IAddFeaturesResult,
+  IDeleteFeaturesResult,
+  IQueryFeaturesResponse
+} from "@esri/arcgis-rest-feature-service";
+
+import { IAnnoFeature } from "./add";
+import { UserSession } from "@esri/arcgis-rest-auth";
+
+export interface IVoteRequestOptions extends IRequestOptions {
+  url: string;
+  annotation: IAnnoFeature;
+  /**
+   * Set to 'true' in order to vote another annotation down.
+   */
+  downVote?: boolean;
+  authentication: UserSession;
+}
+
+/**
+ * ```js
+ * import { vote } from "@esri/hub-annotations";
+ * //
+ * voteOnAnnotation({
+ *   url,
+ *   annotation: { attributes: {} },
+ *   authentication
+ * })
+ *   .then(response);
+ * ```
+ * Upvote (or downvote) another annotation.
+ * @param requestOptions - request options that may include authentication
+ * @returns A Promise that will resolve with the response from the service after attempting to vote on an annotation.
+ */
+export function voteOnAnnotation(
+  requestOptions: IVoteRequestOptions
+): Promise<IAddFeaturesResult | IDeleteFeaturesResult> {
+  const annotation = requestOptions.annotation;
+  const auth = requestOptions.authentication as UserSession;
+  const url = requestOptions.url;
+
+  if (!auth) {
+    return new Promise((resolve, reject) => {
+      reject(
+        new ArcGISAuthError(`Voting by anonymous users is not supported.`)
+      );
+    });
+  }
+
+  if (annotation.attributes.author === auth.username) {
+    return new Promise((resolve, reject) => {
+      reject(
+        new ArcGISRequestError(`Users may not vote on their own comment/idea.`)
+      );
+    });
+  }
+
+  // searchAnnotations() would make more xhrs for user metadata
+  return queryFeatures({
+    url,
+    where: `parent_id = ${annotation.attributes.OBJECTID} AND author = '${
+      auth.username
+    }'`
+  }).then((
+    queryResponse: any /* why can't i declare it as IQueryFeaturesResponse ? */
+  ) => {
+    if (queryResponse.features.length > 0) {
+      // if its a switch vote, call deleteFeatures() to remove the original
+      if (
+        (!requestOptions.downVote &&
+          queryResponse.features[0].attributes.value === -1) ||
+        (requestOptions.downVote &&
+          queryResponse.features[0].attributes.value === 1)
+      ) {
+        const deleteOptions: IDeleteFeaturesRequestOptions = {
+          url,
+          authentication: requestOptions.authentication,
+          objectIds: [queryResponse.features[0].attributes.OBJECTID]
+        };
+        return (
+          deleteFeatures(deleteOptions)
+            // i have NO idea why the union return type is being ignored
+            .then(() => {
+              return { addResults: [{ success: true }] } as IAddFeaturesResult;
```

Comment:
in cc576c880617b57cb64443a0fa30d48f4ff0809f i added support for vote switching. this is accomplished by deleting the previous vote (if the new vote is different than the first).

for reasons i don't understand, i had to shape the response of the deleteFeatures() call to impersonate `IAddFeaturesResult` in order to get TypeScript to shut up.

Reply:
it seems to me that after deleting we still need to add the new vote, but you `return` here, so I don't think that happens. Am I missing something?